core/nakamoto/mempool.go

```diff
@@ -19,16 +30,67 @@ type FeeRates struct {
 
 // NewMempool creates a new mempool.
 func NewMempool() *Mempool {
-	return &Mempool{}
+	return &Mempool{
+		candidateTxns: []*RawTransaction{},
+		confirmedTxns: []*Transaction{},
+		fees: &FeeRates{
+			MinFee:    math.MaxUint64,
+			MedianFee: 0,
+			MaxFee:    0,
+		},
+	}
+}
+
+func (m *Mempool) insertCandidate(tx *RawTransaction) {
+	index := sort.Search(len(m.candidateTxns), func(i int) bool {
+		return m.candidateTxns[i].Fee >= tx.Fee
+	})
+
+	// fmt.Printf("index: %d\n", index)
+
+	newArr := make([]*RawTransaction, len(m.candidateTxns)+1)
+	copy(newArr[:index], m.candidateTxns[:index])
+	newArr[index] = tx
+	copy(newArr[index+1:], m.candidateTxns[index:])
+	m.candidateTxns = newArr
+}
+
+func (m *Mempool) setMedianFee() {
+	length := len(m.candidateTxns)
+	if length == 0 {
+		return
+	}
+	if length%2 == 0 {
+		m.fees.MedianFee = (m.candidateTxns[length/2-1].Fee + m.candidateTxns[length/2].Fee) / 2
+	} else if length%2 == 1 {
+		m.fees.MedianFee = m.candidateTxns[length/2].Fee
+	}
+}
+
+func isValidTx(tx *RawTransaction) bool {
+	validSig := core.VerifySignature(hex.EncodeToString(tx.FromPubkey[:]), tx.Sig[:], tx.Envelope())
```

Comment:
Nice, this makes sense

Reply:
Also FYI we need to prevent tx replay. I haven't built this yet but letting you know for now